### PR TITLE
GPU support for cracking Apple iWork files

### PR DESCRIPTION
**Pull request by kholia**
DO NOT MERGE YET!

@magnumripper how is this code even working when the "currentsalt.length" is set to a wrong value in the "set_salt" function?

### CODE REVIEW

**Comment by magnumripper** [comment]
In line 197 you pass -DSALTLEN=16 to the kernel, which overrides the struct's value.

**Comment by kholia** [comment]
@magnumripper I see! This was really making me nervous, LOL! :)

It is easy to use the "split" kernel, and should I use it?

**Comment by magnumripper** [comment]
You should not introduce any new formats with the unsplit kernel though (on the contrary all such formats should be fixed). Base the format on encfs instead.

**Comment by kholia** [comment]
@magnumripper understood, will do, on it! :+1:

**Comment by magnumripper** [comment]
You also had this from Travis CI
```
Testing: iwork-opencl, Apple iWork '09/ '13 / '14 [PBKDF2-SHA1 OpenCL AES]... FAILED (iwork-opencl sets FMT_CASE but all test-vectors are case-insensitive)
```

**Comment by kholia** [comment]
@magnumripper one more question, the iWork '09 test vector uses a 8-byte salt, so how is that vector passing the self-test (since  -DSALTLEN=16 is fixed)?

**Comment by magnumripper** [comment]
Hmmm I was wrong! SALTLEN only defines the size of salt->salt and the kernel does always use salt->length. I'm not sure why it works...

Anyway you should rebase it on encfs and go from there.

File: src/opencl_iwork_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,497 @@
+/* JtR OpenCL format to crack iWork '13 + '14 files.
+ *
+ * This software is Copyright (c) 2015, Dhiru Kholia <kholia at kth.se> and
+ * Maxime Hulliger <hullger at kth.se>, and it is hereby released to the
```
Comment (magnumripper) [inline src/opencl_iwork_fmt_plug.c]: Is this correct? Hulliger != hullger

**Comment by kholia** [comment]
@magnumripper all the changes are in place now, the new "iwork-opencl" format fails for some unknown reason(s) though.

File: src/iwork_common_plug.c [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		goto err;
+	if(hexlenl(p) != IVLEN * 2)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // blob
+		goto err;
+	if(hexlenl(p) != BLOBLEN * 2)
+		goto err;
+	if ((p = strtokm(NULL, "$")) != NULL)
+		goto err;
+
```
Comment (magnumripper) [inline src/iwork_common_plug.c]: Use a space between `if` and `(` please. Several instances.

File: src/iwork_common_plug.c [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+int iwork_common_decrypt(struct format_context *fctx, unsigned char *key, unsigned char *iv, unsigned char *data)
+{
+	unsigned char out[BLOBLEN] = {0};
+	unsigned char ivec[IVLEN];
+	uint8_t hash[32];
+	SHA256_CTX ctx;
+
+	AES_KEY aes_decrypt_key;
+	AES_set_decrypt_key(key, 128, &aes_decrypt_key);
```
Comment (magnumripper) [inline src/iwork_common_plug.c]: There should be a blank line after `AES_KEY aes_decrypt_key;`, not before it.

File: src/opencl_iwork_fmt_plug.c [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+john_register_one(&fmt_opencl_iwork);
+#else
+
+#include <string.h>
+#include "aes.h"
+#include "hmac_sha.h"
+#ifdef _OPENMP
+#include <omp.h>
+#endif
+
```
Comment (magnumripper) [inline src/opencl_iwork_fmt_plug.c]: Do not mix system headers and local ones. First, include all `<system.h>` then we use to have a blank line, the you list all `"local.h"`

File: src/opencl_iwork_fmt_plug.c [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+#define cracked_size (sizeof(*cracked) * global_work_size * ocl_v_width)
+
+/*
+ * HASH_LOOPS is ideally made by factors of (iteration count - 1) and should
+ * be chosen for a kernel duration of not more than 200 ms
+ */
+#define HASH_LOOPS              (3 * 251)
+#define ITERATIONS              181474 /* Just for auto tune */
+#define LOOP_COUNT              (((currentsalt.iterations - 1 + HASH_LOOPS - 1)) / HASH_LOOPS)
```
Comment (magnumripper) [inline src/opencl_iwork_fmt_plug.c]: HASH_LOOPS should better be `(3 * 271)` for being optimized for 100,000 iterations. And ITERATIONS should be `100000` for better auto-tune.

File: src/opencl_iwork_fmt_plug.c [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+	fctx = (iwork_common_custom_salt*)salt;
+	memcpy((char*)currentsalt.salt, fctx->salt, fctx->salt_length);
+	currentsalt.length = fctx->salt_length;
+	currentsalt.iterations = fctx->iterations;
+	currentsalt.outlen = 16; // AES 128-bit key only
+	HANDLE_CLERROR(clEnqueueWriteBuffer(queue[gpu_id], mem_salt, CL_FALSE, 0, sizeof(pbkdf2_salt), &currentsalt, 0, NULL, NULL), "Copy salt to gpu");
+}
+
+static void clear_keys(void) {
+	memset(inbuffer, 0, key_buf_size);
```
Comment (magnumripper) [inline src/opencl_iwork_fmt_plug.c]:
```diff
- static void clear_keys(void) {
+ static void clear_keys(void)
+ {
```

File: src/opencl_iwork_fmt_plug.c [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+
+	/// Read the result back
+	BENCH_CLERROR(clEnqueueReadBuffer(queue[gpu_id], mem_out, CL_TRUE, 0, sizeof(pbkdf2_out) * scalar_gws, output, 0, NULL, multi_profilingEvent[4]), "Copy result back");
+
+	if (!ocl_autotune_running) {
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+		for (index = 0; index < count; index++)
+		if (iwork_common_decrypt(fctx, (unsigned char*)output[index].dk, fctx->iv, fctx->blob) == 0)
```
Comment (magnumripper) [inline src/opencl_iwork_fmt_plug.c]: Well, if you drop the ` == 0` in the end, it works like a champ :laughing:

**Comment by magnumripper** [comment]
Too easy :wink:

**Comment by kholia** [comment]
LOL :-(

**Comment by kholia** [comment]
@magnumripper this should be in better shape now, thanks! :+1:

**Comment by magnumripper** [comment]
Yeah and it passes the robots. So let's merge it, no?

**Pull request merged**